kong/core/handler.lua

```diff
@@ -118,16 +125,10 @@ return {
           "' with: "..tostring(err))
       end
 
-      -- if set `host_header` is the original header to be preserved
-      var.upstream_host = host_header or 
+      -- if set, `upstream_host` is the original header to be preserved
+      var.upstream_host = ctx.upstream_host or
           balancer_address.hostname..":"..balancer_address.port
 
```

Comment:
The balancer context is too late in the execution phases. It won't take the value from `var.upstream_host` to use it in the upstream header since ngx_http_proxy_module will have ran already. Tested.